File: src/main/java/com/intgrah/algorithms/list/DoublyLinkedList.java
package com.intgrah.algorithms.list;

import java.util.Iterator;

public class DoublyLinkedList<T> extends AbstractList<T> implements
        Deque<T, DoublyLinkedList<T>.Node> {

    Node front;
    Node back;

    @Override
    public T getFront() {
        assert !isEmpty();
        return front.value;
    }

    @Override
    public T popFront() {
        assert !isEmpty();
        Node f = front;
        front = f.next;
        if (--size == 0)
            back = null;
        else
            front.prev = null;
        return f.value;
    }

    @Override
    public Node pushBack(T value) {
        Node n = new Node(value);
        if (size++ == 0) {
            front = n;
        } else {
            n.prev = back;
            back.next = n;
        }
        back = n;
        return n;
    }

    @Override
    public T getBack() {
        assert !isEmpty();
        return back.value;
    }

    @Override
    public T popBack() {
        assert !isEmpty();
        Node b = back;
        back = b.prev;
        if (--size == 0)
            front = null;
        else
            back.next = null;
        return b.value;
    }

    @Override
    public Node pushFront(T value) {
        Node n = new Node(value);
        if (size++ == 0) {
            back = n;
        } else {
            n.next = front;
            front.prev = n;
        }
        front = n;
        return n;
    }

    public void append(DoublyLinkedList<T> l) {
        if (l.size == 0)
            return;
        if (size == 0) {
            front = l.front;
        } else {
            back.next = l.front;
            l.front.prev = back;
        }
        back = l.back;
        size += l.size;
        l.clear();
    }

    @Override
    public void clear() {
        front = back = null;
        size = 0;
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<>() {
            private Node n = front;

            @Override
            public boolean hasNext() { return n != null; }

            @Override
            public T next() {
                T value = n.value;
                n = n.next;
                return value;
            }
        };
    }

    public class Node implements Deletable {

        private final T value;
        private Node prev;
        private Node next;

        Node(T v) { value = v; }

        public void delete() {
            assert !isEmpty();
            if (prev == null)
                front = next;
            else
                prev.next = next;
            if (next == null)
                back = prev;
            else
                next.prev = prev;
            size--;
        }

    }

}

File: list/doubly_linked_list.py
from collections.abc import Iterator


class DoubleNode[T]:
    def __init__(self, value: T):
        self.value = value
        self.next: DoubleNode[T] | None = None
        self.prev: DoubleNode[T] | None = None

class DoublyLinkedList[T]:
    def __init__(self):
        self.front: DoubleNode[T] | None = None
        self.back: DoubleNode[T] | None = None
        self.len = 0

    def push_front(self, value: T) -> DoubleNode[T]:
        n = DoubleNode(value)
        if self.front is None:
            self.front = self.back = n
        else:
            n.next = self.front
            self.front.prev = n
            self.front = n
        self.len += 1
        return n

    def pop_front(self) -> T:
        if self.front is None:
            raise IndexError
        f = self.front
        self.front = f.next
        self.front.prev = None
        self.len -= 1
        return f.value
    
    def push_back(self, value: T) -> DoubleNode[T]:
        n = DoubleNode(value)
        if self.back is None:
            self.front = self.back = n
        else:
            n.prev = self.back
            self.back.next = n
            self.back = n
        self.len += 1
        return n

    def pop_back(self) -> T:
        b = self.back
        self.back = b.prev
        self.back.next = None
        self.len -= 1
        return b.value
    
    def delete(self, n: DoubleNode[T]) -> None:
        if n.prev is None:
            self.front = n.next
        else:
            n.prev.next = n.next
        if n.next is None:
            self.back = n.prev
        else:
            n.next.prev = n.prev
        self.len -= 1
            
    def clear(self) -> None:
        self.front = self.back = None
        self.len = 0
    
    def __iter__(self) -> Iterator[T]:
        n = self.front
        while n is not None:
            yield n.value
            n = n.next

    def __len__(self) -> int:
        return self.len
